internal/discoverer/zookeeper.go

```diff
@@ -0,0 +1,297 @@
+package discoverer
+
+import (
+	"encoding/json"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/api7/apisix-seed/internal/conf"
+	"github.com/api7/apisix-seed/internal/core/comm"
+	"github.com/api7/apisix-seed/internal/log"
+	"github.com/api7/apisix-seed/internal/utils"
+	"github.com/go-zookeeper/zk"
+	"golang.org/x/net/context"
+)
+
+func init() {
+	Discoveries["zookeeper"] = NewZookeeperDiscoverer
+}
+
+type ZookeeperService struct {
+	Name         string
+	BindEntities []string
+	WatchPath    string
+	WatchContext context.Context
+	WatchCancel  context.CancelFunc
+}
+
+type ZookeeperDiscoverer struct {
+	zkConfig          *conf.Zookeeper
+	zkConn            *zk.Conn
+	zkWatchServices   sync.Map
+	zkUnWatchServices sync.Map
+	zkUnWatchContext  context.Context
+	zkUnWatchCancel   context.CancelFunc
+
+	msgCh chan *comm.Message
+}
+
+func (zd *ZookeeperDiscoverer) Stop() {
+	zd.zkWatchServices.Range(func(key, value interface{}) bool {
+		zd.removeWatchService(value.(*ZookeeperService))
+		return true
+	})
+	close(zd.msgCh)
+	zd.zkConn.Close()
+	zd.zkUnWatchCancel()
+}
+
+func (zd *ZookeeperDiscoverer) Query(query *comm.Query) error {
+	values, _, err := query.Decode()
+	if err != nil {
+		return err
+	}
+
+	event, entity, serviceName := values[0], values[1], values[2]
+
+	switch event {
+	case utils.EventAdd:
+		err = zd.fetchService(serviceName, []string{entity})
+	case utils.EventDelete:
+		err = zd.removeService(serviceName)
+	}
+
+	return err
+}
+
+func (zd *ZookeeperDiscoverer) Update(update *comm.Update) error {
+	// TODO:
+	// Zookeeper does not need to introduce parameter information like nacos,
+	// so the Update interface cannot be triggered on the framework.
+	return nil
+}
+
+func (zd *ZookeeperDiscoverer) Watch() chan *comm.Message {
+	return zd.msgCh
+}
+
+// fetchService fetch service watch and send message notify
+func (zd *ZookeeperDiscoverer) fetchService(serviceName string, entities []string) error {
+	zkService, ok := zd.zkWatchServices.Load(serviceName)
```

Comment:
We can do the `zkService.(*ZookeeperService)` conversion here?

Reply:
We can use another var as an alias?

Reply:
done